Hack scripts to install demos in OpenShift

josunect opened the pull request:
Adding some hack scripts to install demos in OpenShift.

The idea is they work the same - Having each one their own particularities: 

- h for help 
- d for uninstall 
- c for client (oc by default, tested also for kubectl)

demos added: 

- electronic shop
- federated travels: When installing this one in open shift, for some reason, it doesn't look to be injecting pro pertly the Istio injections, but it looks to be installed in the mesh. To be reviewed. 
- music store
- runtimes
- scale mesh
- service spawner: For some reason I'm not able to install the example. There is a crashLoopBackOff error trying to deploy the image quay.io/jotak/nginx-hello:0.2 - I should look into this, or probably remove the script in the meantime. 
- topology generator

Fixes https://github.com/kiali/kiali/issues/5268

jmazzitelli commented:
> * service spawner: For some reason I'm not able to install the example. There is a crashLoopBackOff error trying to deploy the image quay.io/jotak/nginx-hello:0.2 - I should look into this, or probably remove the script in the meantime.

"jotak" is @jotak -- he wrote that stuff back when he was on the Kiali team. He can tell you what is going on.

josunect commented:
> > * service spawner: For some reason I'm not able to install the example. There is a crashLoopBackOff error trying to deploy the image quay.io/jotak/nginx-hello:0.2 - I should look into this, or probably remove the script in the meantime.
> 
> "jotak" is @jotak -- he wrote that stuff back when he was on the Kiali team. He can tell you what is going on.

Thank you @jmazzitelli , I'll ask him.

jmazzitelli commented:
I installed the music-store demo and didn't see a kiali graph. Probably because this new hack script doesn't fully install the demo as documented here: https://github.com/kiali/demos/tree/master/music-store#quick-start

We should print something out in the help to tell the person to finish the install using the docs spelled out there.

Or you could finish the install in the script itself.

But right now, it is confusing because the install script finished successfully but I don't see anything in the Kiali graph and it looks like it isn't working.

jmazzitelli commented on hack/istio/install-electronic-shop-demo.sh:
```diff
@@ -0,0 +1,110 @@
+else
+
+    echo "Deleting the '${ESHOP}' app in the '${ESHOP}' namespace..."
+    ${CLIENT_EXE} delete -n ${ESHOP} -f <(curl -L ${BASE_RUL}/${ESHOP}/${ESHOP}.yaml)
```
Bad var name. Consider using "set -u" at the top of these scripts to fail-fast if you attempt to use an undefined variable.

Right now this fails when passing `-d true` to delete the demo:

```
$ ./install-electronic-shop-demo.sh -d true
CLIENT_EXE=oc
IS_OPENSHIFT=true
Deleting the 'electronic-shop' app in the 'electronic-shop' namespace...
curl: (3) <url> malformed
```
```suggestion
    ${CLIENT_EXE} delete -n ${ESHOP} -f <(curl -L ${BASE_URL}/${ESHOP}/${ESHOP}.yaml)
```

jmazzitelli commented on hack/istio/install-topology-generator-demo.sh:
```diff
@@ -0,0 +1,147 @@
+      cat <<HELPMSG
+Valid command line arguments:
+  -c|--client: either 'oc' or 'kubectl'
+  -n|--num_ns: number of namespaces to be generate in the generator to apply the network policies
```
This is minor but two things here:
1) all of our other options in all our hack scripts do not use `_` (underscore), use a `-` (dash) between words in option names.
2) To be consistent, consider making this `--namespaces` just like you have for the scale-mesh script.

So, either rename this `--num-ns` (with dash) or `--namespaces` (for consistency with the other scripts). e.g.

```suggestion
  -n|--num-ns: number of namespaces to be generate in the generator to apply the network policies
```

or

```suggestion
  -n|--namespaces: number of namespaces to be generate in the generator to apply the network policies
```

jmazzitelli commented on hack/istio/install-music-store-demo.sh:
```diff
@@ -0,0 +1,173 @@
+  ${CLIENT_EXE} wait --timeout 60s --for condition=available deployment/music-store-ui-v1 -n music-store
+
+  export INGRESS_PORT=$(${CLIENT_EXE} -n istio-system get service istio-ingressgateway -o jsonpath='{.spec.ports[?(@.name=="http2")].nodePort}')
+  if [ "${IS_OPENSHIFT}" == "true" ]; then
+       export INGRESS_HOST=$(crc ip)
+    else
+       export INGRESS_HOST=$(minikube ip)
+  fi
```
`else` is indented in a different column from if/fi - this doesn't really matter in bash but it is confusing to read because the alignment is weird. Also, some of this is indented with 4 or 5 spaces, other times in these scripts the indentations are 2 spaces. Make your indentation consistent for better readability and debugging (harder to tell what code is where when the indentations are different throughout the code).

Here's suggestion for this portion, but I suggest going through these scripts and making the indentation consistent.

```suggestion
  if [ "${IS_OPENSHIFT}" == "true" ]; then
    export INGRESS_HOST=$(crc ip)
  else
    export INGRESS_HOST=$(minikube ip)
  fi
```

jmazzitelli commented on hack/istio/install-topology-generator-demo.sh:
```diff
@@ -0,0 +1,147 @@
+      NUM_NS="$2"
+      shift;shift
+      ;;
+    -g|--generate_config)
```
here's another one with `_` - just make it `-`:

```suggestion
    -g|--generate-config)
```

jmazzitelli commented on hack/istio/install-topology-generator-demo.sh:
```diff
@@ -0,0 +1,147 @@
+  -c|--client: either 'oc' or 'kubectl'
+  -n|--num_ns: number of namespaces to be generate in the generator to apply the network policies
+  -d|--delete: if 'true' demos will be deleted; otherwise, they will be installed
+  -g|--generate_config: Generate configuration of the topology based in the number of namespaces (-n)
```
use dash:
```suggestion
  -g|--generate-config: Generate configuration of the topology based in the number of namespaces (-n)
```

jmazzitelli commented on hack/istio/install-topology-generator-demo.sh:
```diff
@@ -0,0 +1,147 @@
+  1) Run with no args - or -c - to install the topology generator
+  2) Run the proxy: CLIENT port-forward svc/topology-generator 8080:8080 -n topology-generator
+  3) Go to the app and generate the topology
+  4) Run -g to run the openshift configuration and -n with the number of namespaces created
```
I was confused here and had to look at the code to see what this is doing. It isn't clear what this "generate-config" option does. I think we should make the help text more clear.

IIUC, this option is only for OpenShift AND upstream Istio - if you are on Minikube or other non-OpenShift cluster, you do not need this. If you are using OSSM, you do not need this. But if you are on OpenShift using upstream Istio, this `generate-config` will create the necessary NAD and SCCs that are required.

So perhaps provide a little more detail here to help the user understand what this option is and when it is needed. I had to read the code to figure it out.

Something like:

`4) Run -g to configure the demo to work on OpenShift. You only need to run this if you are on OpenShift and using upstream Istio (not OSSM).`

josunect replied:
Updated help to clarify this behavior

josunect commented:
> I installed the music-store demo and didn't see a kiali graph. Probably because this new hack script doesn't fully install the demo as documented here: https://github.com/kiali/demos/tree/master/music-store#quick-start
> 
> We should print something out in the help to tell the person to finish the install using the docs spelled out there.
> 
> Or you could finish the install in the script itself.
> 
> But right now, it is confusing because the install script finished successfully but I don't see anything in the Kiali graph and it looks like it isn't working.

I've tested and changed a couple of things because It looks like it was not applying correctly the security context. 
I have changed the message because the only thing the script is not doing is generate traffic, so after the installation you can see the message:

You should be able to access the API with the url: http://192.168.130.11:31788/api in order to generate traffic
You should be able to access the UI with the url: http://192.168.130.11:31788/ in order to generate traffic

So, after hitting the url, it should be possible to see the traffic in Kiali:

![image](https://user-images.githubusercontent.com/49480155/177123640-91e8ca5b-df25-4e0f-9da3-ece613c497ac.png)

josunect commented:
> > > * service spawner: For some reason I'm not able to install the example. There is a crashLoopBackOff error trying to deploy the image quay.io/jotak/nginx-hello:0.2 - I should look into this, or probably remove the script in the meantime.
> > 
> > 
> > "jotak" is @jotak -- he wrote that stuff back when he was on the Kiali team. He can tell you what is going on.
> 
> Thank you @jmazzitelli , I'll ask him.

The demo service-spawner works in minikube but not in OpenShift. 
Running just the image that is failing:

![image](https://user-images.githubusercontent.com/49480155/177162744-7ce4ffde-9955-4f30-94c7-0cc15038abbc.png)

Running just the image we can see a permissions error: 

![image](https://user-images.githubusercontent.com/49480155/177163071-05db6036-f622-4cf6-977f-05a7f56f6c08.png)

So, I guess we can just use the nginxdemos/nginx-hello for the OpenShift demo installation.

jmazzitelli commented on hack/istio/install-topology-generator-demo.sh:
```diff
@@ -0,0 +1,147 @@
+SCC
+}
+
+install_topology_generator_demo() {
+
+    if [ "${IS_OPENSHIFT}" == "true" ]; then
+        ${CLIENT_EXE} new-project ${TOPO}
+      else
+        ${CLIENT_EXE} create ns ${TOPO}
+    fi
+    ${CLIENT_EXE} label namespace  ${TOPO} istio-injection=enabled --overwrite=true
+
+    if [ "${IS_OPENSHIFT}" == "true" ]; then
+      apply_network_attachment ${TOPO}
+    fi
+
+    ${CLIENT_EXE} apply -n ${TOPO}  -f ${BASE_URL}/topology-generator/deploy/generator.yaml
+}
+
+generate_config() {
+
+  if [ "${IS_OPENSHIFT}" == "true" ]; then
+    for (( i=1; i<=$NUM_NS; i++ ))
+        do
+
+            apply_network_attachment n${i}
+    done
+  fi
+}
```
Just reformat - the indentation was haphazard:

```suggestion
install_topology_generator_demo() {
  if [ "${IS_OPENSHIFT}" == "true" ]; then
    ${CLIENT_EXE} new-project ${TOPO}
  else
    ${CLIENT_EXE} create ns ${TOPO}
  fi
  ${CLIENT_EXE} label namespace  ${TOPO} istio-injection=enabled --overwrite=true

  if [ "${IS_OPENSHIFT}" == "true" ]; then
    apply_network_attachment ${TOPO}
  fi

  ${CLIENT_EXE} apply -n ${TOPO}  -f ${BASE_URL}/topology-generator/deploy/generator.yaml
}

generate_config() {
  if [ "${IS_OPENSHIFT}" == "true" ]; then
    for (( i=1; i<=$NUM_NS; i++ ))
    do
      apply_network_attachment n${i}
    done
  fi
}
```

jmazzitelli commented on hack/istio/install-music-store-demo.sh:
```diff
@@ -0,0 +1,169 @@
+  if [ "${IS_OPENSHIFT}" == "true" ]; then
+    export INGRESS_HOST=$(crc ip)
+  else
+    export INGRESS_HOST=$(minikube ip)
```
This line breaks for me because my minikube profile is not the default `minikube`:

```
You should be able to access the API with the url: http://🤷  Profile "minikube" not found. Run "minikube profile list" to view all profiles.
👉  To start a cluster, run: "minikube start":30755/api in order to generate traffic
You should be able to access the UI with the url: http://🤷  Profile "minikube" not found. Run "minikube profile list" to view all profiles.
👉  To start a cluster, run: "minikube start":30755/ in order to generate traffic
```
We should allow the user to pass in an option `--minikube-profile` (`--mp`) and use that (its default should be `minikube`). We do this in other hack scripts so you can just copy what we do there. For example, in the bookinfo demo install hack script see: https://github.com/kiali/kiali/blob/master/hack/istio/install-bookinfo-demo.sh#L82-L85 and https://github.com/kiali/kiali/blob/master/hack/istio/install-bookinfo-demo.sh#L308-L309

Once you have a `--minikube-profile` option that optionally sets MINIKUBE_PROFILE, you can have this line be:
```suggestion
    export INGRESS_HOST=$(minikube -p ${MINIKUBE_PROFILE:-minikube} ip)
```

jmazzitelli reviewed (changes requested):
a couple requested changes:

1) update "minikube ip" to accept a profile for people like me who run multiple different minikube clusters with different profile names.
2) trivial - fix the formatting of a block of code for consistency

jmazzitelli commented on hack/istio/install-music-store-demo.sh:
```diff
@@ -12,6 +12,8 @@
 : ${DELETE_DEMOS:=false}
 : ${MSTORE:=music-store}
 : ${BASE_URL:=https://raw.githubusercontent.com/kiali/demos/master}
+: ${MINIKUBE_PROFILE=minikube}
+: ${ISTIO_NAMESPACE=istio-system}
```
The semantics for `=` is different than `:=` (in fact, I don't know what the rule is if using just `=` here).
So I think you need a `:` character in here - bash has very different semantics depending on how the value is declared here - read this for the gory details: https://tiswww.case.edu/php/chet/bash/bashref.html#Shell-Parameter-Expansion

So you want this:

```suggestion
: ${MINIKUBE_PROFILE:=minikube}
: ${ISTIO_NAMESPACE:=istio-system}
```

josunect replied:
Thanks, @jmazzitelli !

jmazzitelli reviewed (changes requested):
other than the missing `:` in the env var assignments, this should be good.

jmazzitelli reviewed (approved):
LGTM
